### PR TITLE
Fixes from scala/scala repository

### LINKED ISSUE #214: Backport work that has been made on scala/scala
For instance: https://github.com/scala/scala/pull/5972

Here is the history of the collections: https://github.com/scala/scala/commits/2.13.x/src/library/scala/collection
Here is the history of the JUnit tests: https://github.com/scala/scala/commits/2.13.x/test/junit/scala/collection

### PR DESCRIPTION
**Pull request by julienrf**
I went through the history of the collections on the scala/scala repo, for the last year, and backported all the things that were missing in the strawman.

Fixes #214.

### CODE REVIEW

**Comment by julienrf** [comment]
I’m wondering what happened to https://github.com/scala/scala/pull/5664, it seems to have been reverted (https://github.com/scala/scala/commit/a11918d1053f3b0af32d7bd7fb5b75acb31724ff, https://github.com/scala/scala/commit/76babbb6728e5daf4fb95444273f6f8e0a4098d7 and https://github.com/scala/scala/commit/0365d58f62e5f21a2723365cb21636f382474805). Also, it seems that we have a similar PR (#354). What should we do about this stuff?

**Comment by lrytz** [comment]
https://github.com/scala/scala/pull/5664 was partially reverted because it was binary incompatible. We don't need to worry about that here.

**Comment by szeiger** [comment]
#354 is the intended replacement for the parts in https://github.com/scala/scala/pull/5664 that had to be reverted due to binary compatibility constraints.

File: collections/src/main/scala/strawman/collection/LinearSeq.scala [hunk 1]
```diff
@@ -171,4 +171,8 @@ trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A]] extends Any w
     }
     last
   }
+
+  override def tails: Iterator[C] =
+    Iterator.iterate(coll)(_.tail.asInstanceOf[C]).takeWhile(_.nonEmpty) ++ Iterator(newSpecificBuilder().result())
```
Comment (szeiger) [inline collections/src/main/scala/strawman/collection/LinearSeq.scala]: Is the cast necessary?

File: collections/src/main/scala/strawman/collection/package.scala [hunk 2]
```diff
@@ -121,8 +121,8 @@ package object collection extends LowPriority {
 
   /** An extractor used to init/last deconstruct sequences. */
   object :+ {
-    /** Splits a sequence into init :+ last.
-      * @return Some((init, last)) if sequence is non-empty. None otherwise.
+    /** Splits a sequence into init :+ tail.
+      * @return Some((init, tail)) if sequence is non-empty. None otherwise.
```
Comment (szeiger) [inline collections/src/main/scala/strawman/collection/package.scala]: This is wrong. `:+` does split the sequence into `(init, last)`.

File: collections/src/main/scala/strawman/collection/Iterable.scala [hunk 3]
```diff
@@ -361,6 +361,13 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     */
   def splitAt(n: Int): (C, C) = (take(n), drop(n))
 
+  /** Selects first ''n'' elements.
+    *  $orderDependent
+    *  @param  n    the number of elements to take from this $coll.
+    *  @return a $coll consisting only of the first `n` elements of this $coll,
+    *          or else the whole $coll, if it has less than `n` elements.
+    *          If `n` is negative, returns an empty $coll.
+    */
```
[the remark this answers is not in the document]
Comment (julienrf) [reply collections/src/main/scala/strawman/collection/Iterable.scala]: I’ve put them here to override `IterableOnce`’s comments which give information about the fact that these operation consume the `Iterator`. This information is not relevant in `Iterable`.

But maybe we could organize the scaladoc differently…

File: collections/src/main/scala/strawman/collection/LinearSeq.scala (hunk 1 above)
Comment (julienrf) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: Yes, the static return type of `tail` is `LinearSeq[A]`, whereas we want a `C`.

File: collections/src/main/scala/strawman/collection/package.scala (hunk 2 above)
Comment (julienrf) [reply collections/src/main/scala/strawman/collection/package.scala]: Oops, indeed…

**Comment by julienrf** [comment]
> #354 is the intended replacement for the parts in scala/scala#5664 that had to be reverted due to binary compatibility constraints.

OK, so that’s fine to not have it in this PR.

File: collections/src/main/scala/strawman/collection/Iterable.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (szeiger) [reply collections/src/main/scala/strawman/collection/Iterable.scala]: The iterator comsumption stuff is contained in templates (or whatever you call those scaladoc variables). They should be defined as empty in `Iterable` and with the relevant text in `Iterator`.

File: collections/src/main/scala/strawman/collection/LinearSeq.scala (hunk 1 above)
Comment (szeiger) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: Hm, that's not nice. We lose the precise type on repeated applications of `tail`. `this.type.tail` results in C but `C.type` is only `LinearSeq[A]`.

Comment (julienrf) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: Having a more precise type would require a more complex upper bound on `C`.

Comment (szeiger) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: The problem lies in the definition of `LinearSeqOps`. We need to make `C` recursive:
```scala
trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A] with LinearSeqOps[A, CC, C]] extends Any with SeqOps[A, CC, C]
```
And the same for `immutable.LinearSeqOps` and `LazyListOps`. With these changes the cast is no longer needed.

Comment (julienrf) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: Yes, that’s what I was trying to say in my previous comment. The drawback is that it becomes harder to abstract over _some_ `LinearSeqOps`. See how it is done for `immutable.MapOps`, which has similar bounds: https://github.com/scala/collection-strawman/blob/master/collections-contrib/src/main/scala/strawman/collection/decorators/HasImmutableMapOps.scala#L12-L17

Comment (szeiger) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: I don't think this is a problem for `LinearSeqOps`. The only changes I had to do downstream was to repeat the bounds in `immutable.LinearSeqOps` and `LazyListOps`. `LinearSeqOps` is just an implementation detail. You'd abstract over `SeqOps`, possibly restricted to subtypes of `LinearSeq` if you care about performance characteristics.

File: collections/src/main/scala/strawman/collection/Iterable.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (julienrf) [reply collections/src/main/scala/strawman/collection/Iterable.scala]: In ab2d3f38c5f2357dc23829d236e5c38e795de797 I moved the documentation to `IterableOnceOps` but there is still an issue with the documentation of `IterableOnceOps`. I’ve opened #510 to keep track of it.

File: collections/src/main/scala/strawman/collection/LinearSeq.scala (hunk 1 above)
Comment (julienrf) [reply collections/src/main/scala/strawman/collection/LinearSeq.scala]: Allright, I’ve fixed that in a304aaf.

**Pull request merged**